projects/chaos/src/power_path.c

```diff
@@ -0,0 +1,147 @@
+    event_raise(CHAOS_EVENT_CAN_UV_OV, POWER_PATH_SOURCE_ID_AUX_BAT);
+  }
+}
+
+static void prv_adc_read(SoftTimerID timer_id, void *context) {
+  PowerPathSource *pps = context;
+  if (pps->timer_id != timer_id || !pps->monitoring_active) {
+    // Guard against accidentally starting multiple timers to monitor the same source concurrently.
+    return;
+  }
+
+  uint16_t value = 0;
+  ADCChannel chan = NUM_ADC_CHANNELS;
+  // Read and convert the current values.
+  adc_get_channel(pps->current_pin, &chan);
+  adc_read_raw(chan, &value);
+  pps->readings.current = pps->current_convert(value);
+
+  // Read and convert the voltage values.
+  adc_get_channel(pps->voltage_pin, &chan);
+  adc_read_raw(chan, &value);
+  pps->readings.voltage = pps->voltage_convert(value);
+
+  // Start the next timer.
+  soft_timer_start(pps->period_us, prv_adc_read, pps, &pps->timer_id);
+}
+
+StatusCode power_path_init(PowerPathCfg *pp) {
+  GPIOSettings settings = {
+    .direction = GPIO_DIR_OUT,
+    .state = GPIO_STATE_HIGH,
+    .resistor = GPIO_RES_NONE,
+    .alt_function = GPIO_ALTFN_NONE,
+  };
+
+  status_ok_or_return(gpio_init_pin(&pp->enable_pin, &settings));
+  status_ok_or_return(gpio_init_pin(&pp->shutdown_pin, &settings));
+
+  // Configure the Interrupt Pins
+  settings.direction = GPIO_DIR_IN;
+  settings.state = GPIO_STATE_LOW;
+  status_ok_or_return(gpio_init_pin(&pp->aux_bat.uv_ov_pin, &settings));
+  status_ok_or_return(gpio_init_pin(&pp->dcdc.uv_ov_pin, &settings));
+
+  // Register interrupts to the same function.
+  const InterruptSettings it_settings = {
+    .type = INTERRUPT_TYPE_INTERRUPT,
+    .priority = INTERRUPT_PRIORITY_NORMAL,
+  };
+  status_ok_or_return(gpio_it_register_interrupt(&pp->aux_bat.uv_ov_pin, &it_settings,
+                                                 INTERRUPT_EDGE_RISING, prv_voltage_warning, pp));
+  status_ok_or_return(gpio_it_register_interrupt(&pp->dcdc.uv_ov_pin, &it_settings,
+                                                 INTERRUPT_EDGE_RISING, prv_voltage_warning, pp));
+
+  // Configure the ADC Pins
+  settings.alt_function = GPIO_ALTFN_ANALOG;
+  status_ok_or_return(gpio_init_pin(&pp->aux_bat.voltage_pin, &settings));
+  status_ok_or_return(gpio_init_pin(&pp->aux_bat.current_pin, &settings));
+  status_ok_or_return(gpio_init_pin(&pp->dcdc.voltage_pin, &settings));
+  return gpio_init_pin(&pp->dcdc.current_pin, &settings);
+}
+
+StatusCode power_path_enable(const PowerPathCfg *pp) {
+  status_ok_or_return(gpio_set_state(&pp->shutdown_pin, GPIO_STATE_HIGH));
+  return gpio_set_state(&pp->enable_pin, GPIO_STATE_HIGH);
+}
+
+StatusCode power_path_disable(PowerPathCfg *pp) {
+  // Stop interrupts from firing first to avoid false readings after shutdown.
+  status_ok_or_return(power_path_source_monitor_disable(&pp->aux_bat));
+  status_ok_or_return(power_path_source_monitor_disable(&pp->dcdc));
+  status_ok_or_return(gpio_set_state(&pp->shutdown_pin, GPIO_STATE_LOW));
```

Comment:
Wouldn't disabling the powerpath cause the MCU to die?

Reply:
Yes I believe that is correct. I have no intention of actually using this command anywhere but since the hardware is configured to allow to be shutdown I just added this anyway. I'm happy to remove/comment this out if you are concerned it might be used by accident.

Reply:
Yeah, I'm not sure that there's value in disabling the powerpath and it's just more code to maintain. :)

Reply:
Done. Removed.